File: backend/app/schemas/tree.py
from pydantic import BaseModel, ConfigDict, Field
from datetime import datetime
from typing import Optional, List, Dict


class TreeNode(BaseModel):
    id: int
    first_name: str
    last_name: str
    patronymic: Optional[str] = None
    birth_date: Optional[str] = None
    death_date: Optional[str] = None
    gender: Optional[str] = None
    photo_url: Optional[str] = None
    is_active: bool
    is_admin: bool
    depth: int = 0
    user_id: Optional[int] = None

    model_config = ConfigDict(from_attributes=True)


class TreeEdge(BaseModel):
    from_id: int = Field(..., alias="from")
    to_id: int = Field(..., alias="to")
    type: str

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class TreeResponse(BaseModel):
    nodes: List[TreeNode]
    edges: List[TreeEdge]
    root_id: Optional[int]
    family_id: int

    model_config = ConfigDict(from_attributes=True)


class TreeSettings(BaseModel):
    root_member_id: Optional[int] = None
    include_inactive: bool = False
    max_depth: int = Field(10, ge=1, le=20)
    show_dates: bool = True
    show_photos: bool = True

    model_config = ConfigDict(from_attributes=True)


class TreeSearchQuery(BaseModel):
    query: Optional[str] = Field(None, description="Поиск по имени/фамилии")
    search_inactive: bool = False
    birth_date_from: Optional[datetime] = None
    birth_date_to: Optional[datetime] = None


class TreeSearchResult(BaseModel):
    id: int
    first_name: str
    last_name: str
    patronymic: Optional[str] = None
    birth_date: Optional[str] = None
    death_date: Optional[str] = None
    gender: Optional[str] = None
    is_active: bool
    user_id: Optional[int] = None

    model_config = ConfigDict(from_attributes=True)


class SubtreeRequest(BaseModel):
    member_id: int
    direction: str = Field("both", pattern="^(up|down|both)$")
    generations: int = Field(3, ge=1, le=10)


class RootInfo(BaseModel):
    id: int
    name: str
    birth_date: Optional[str] = None
    is_admin: bool

    model_config = ConfigDict(from_attributes=True)



class PDFMetadata(BaseModel):
    family_name: str
    generated_at: str
    total_members: int
    root_person: Optional[TreeNode] = None
    include_contacts: bool = False


class TreeStatistics(BaseModel):
    total_members: int
    active_members: int
    inactive_members: int
    by_gender: Dict[str, int]
    oldest_member_id: Optional[int] = None
    youngest_member_id: Optional[int] = None


class GenerationGroup(BaseModel):
    level: int
    title: str
    members: List[TreeNode]


class PDFExportData(BaseModel):
    """Данные для генерации PDF на фронтенде"""
    meta: PDFMetadata
    statistics: TreeStatistics
    generations: List[GenerationGroup]
    members_flat: List[TreeNode]
    relationships: List[TreeEdge]

    model_config = ConfigDict(from_attributes=True)


class ExportFormat(BaseModel):
    """Параметры экспорта"""
    format: str = Field("pdf", pattern="^(pdf|json|gedcom)$")
    root_member_id: Optional[int] = None
    include_inactive: bool = False
    include_contacts: bool = Field(False, description="Включить контактные данные (только для PDF)")
    include_statistics: bool = True

File: frontend/src/api/tree.ts
import { apiClient } from './client';

export interface TreeNode {
  id: number;
  first_name: string;
  last_name: string;
  patronymic: string | null;
  birth_date: string | null;
  death_date: string | null;
  gender: 'male' | 'female' | null;
  is_active: boolean;
  is_admin: boolean;
  generation: number;
  partners: number[];
  user_id: number | null;
}

export interface TreeEdge {
  from: number;
  to: number;
  type: string;
}

export interface FamilyUnit {
  id: string;
  parents: number[];
  children: number[];
  type: 'nuclear_family' | 'single_parent';
}

export interface TreeResponse {
  nodes: TreeNode[];
  edges: TreeEdge[];
  family_units: FamilyUnit[];
  root_id: number;
  family_id: number;
}

export interface RootInfo {
  id: number;
  name: string;
  birth_date: string | null;
  is_admin: boolean;
}

export interface SearchResult {
  id: number;
  first_name: string;
  last_name: string;
  patronymic: string | null;
  birth_date: string | null;
  death_date: string | null;
  gender: 'male' | 'female' | null;
  is_active: boolean;
  user_id: number | null;
}

export interface RelativesGroup {
  parents: RelativeInfo[];
  children: RelativeInfo[];
  spouses: RelativeInfo[];
  siblings: RelativeInfo[];
}

export interface RelativeInfo {
  id: number;
  first_name: string;
  last_name: string;
  patronymic: string | null;
  gender?: 'male' | 'female' | null;
  relationship_type?: string;
}

export interface SubtreeResponse {
  nodes: TreeNode[];
  edges: TreeEdge[];
  center_id: number;
}

export const treeAPI = {
  /**
   * Получить полное семейное древо
   * @param familyId - ID семьи
   * @param rootMemberId - ID корневого члена 
   * @param includeInactive - показывать неактивных
   * @param maxDepth - глубина дерева 
   */
  getTree: (familyId: number, rootMemberId?: number, includeInactive: boolean = false, maxDepth: number = 10) => {
    const params = new URLSearchParams();
    if (rootMemberId) params.append('root_member_id', String(rootMemberId));
    params.append('include_inactive', String(includeInactive));
    params.append('max_depth', String(maxDepth));
    
    return apiClient.get<TreeResponse>(`/family/${familyId}/tree?${params.toString()}`);
  },

  
   
  getRoots: (familyId: number) =>
    apiClient.get<RootInfo[]>(`/family/${familyId}/tree/roots`),

  /**
   * Поиск по семейному древу
   * @param familyId - ID семьи
   * @param query - поисковый запрос
   * @param searchInactive - искать среди неактивных
   * @param birthDateFrom - дата рождения от
   * @param birthDateTo - дата рождения до
   */
  search: (
    familyId: number,
    query?: string,
    searchInactive: boolean = false,
    birthDateFrom?: string,
    birthDateTo?: string
  ) => {
    const params = new URLSearchParams();
    if (query) params.append('query', query);
    params.append('search_inactive', String(searchInactive));
    if (birthDateFrom) params.append('birth_date_from', birthDateFrom);
    if (birthDateTo) params.append('birth_date_to', birthDateTo);
    
    return apiClient.get<SearchResult[]>(`/family/${familyId}/tree/search?${params.toString()}`);
  },

  getSubtree: (familyId: number, memberId: number, direction: 'up' | 'down' | 'both' = 'both', generations: number = 3) =>
    apiClient.get<SubtreeResponse>(
      `/family/${familyId}/tree/member/${memberId}/subtree?direction=${direction}&generations=${generations}`
    ),

  getRelatives: (familyId: number, memberId: number) =>
    apiClient.get<RelativesGroup>(`/family/${familyId}/tree/member/${memberId}/relatives`),


  deleteRelationship: (familyId: number, relationshipId: number) =>
    apiClient.delete(`/family/${familyId}/tree/relationship/${relationshipId}`),
};

export default treeAPI;